File: 0647-palindromic-substrings/0647-palindromic-substrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        n = len(s)
        dp = [[0 for _ in range(n)] for _ in range(n)]

        res = 0
        for i in range(n):
            dp[i][i] = 1
            res += 1
            if i + 1 < n and s[i] == s[i+1]:
                dp[i][i+1] = 1 
                res += 1
        
        for diff in range(2, n):
            for i in range(n - diff):
                j = i + diff
                if s[i] == s[j] and dp[i+1][j-1]:
                    dp[i][j] = 1
                    res += 1
                    
        return res

File: 0647-palindromic-substrings/0647-palindromic-substrings.go
func countSubstrings(s string) int {
    checkPalindrome := func(i, j int) int {
        cur := 0
            for i >= 0 && j < len(s) && s[i] == s[j] {
                cur += 1
                i--
                j++
            }

        return cur
    }

    res := 0
    for i := 0; i < len(s) - 1; i++ {
        res += checkPalindrome(i, i)
        res += checkPalindrome(i, i+1)
    }

    // don't forget the last character
    return res + 1
}